streams/src/main/java/org/apache/kafka/streams/state/internals/WrappedStateStore.java

```diff
@@ -19,81 +19,66 @@
 import org.apache.kafka.streams.errors.InvalidStateStoreException;
 import org.apache.kafka.streams.processor.ProcessorContext;
 import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.state.TimestampedBytesStore;
 
 /**
  * A storage engine wrapper for utilities like logging, caching, and metering.
  */
-public interface WrappedStateStore extends StateStore {
-
-    /**
-     * Return the inner most storage engine
-     *
-     * @return wrapped inner storage engine
-     */
-    StateStore inner();
-
-    /**
-     * Return the state store this store directly wraps
-     * @return the state store this store directly wraps
-     */
-    StateStore wrappedStore();
-
-    abstract class AbstractStateStore implements WrappedStateStore {
-        final StateStore innerState;
-
-        protected AbstractStateStore(final StateStore inner) {
-            this.innerState = inner;
+public class WrappedStateStore<S extends StateStore> implements StateStore {
+    public static boolean isTimestamped(StateStore stateStore) {
+        if (stateStore instanceof TimestampedBytesStore) {
+            return true;
+        } else if (stateStore instanceof WrappedStateStore) {
+            return isTimestamped(((WrappedStateStore) stateStore).wrappedStore());
+        } else {
+            return false;
         }
+    }
 
-        @Override
-        public void init(final ProcessorContext context,
-                         final StateStore root) {
-            innerState.init(context, root);
-        }
+    private final S innerState;
```

Comment:
nit: either `inner` or `innerStore` or `innerStateStore` (`innerState` is a weird name)

Or maybe even better `wrapped` instead of `inner` ?

Reply:
I think this should be `protected` to avoid calls to `wrappedStore()` in all child classes

Reply:
I can rename it.

Making it anything but private would expose the field to the whole rest of the `state.internals` package. Not sure why we want to avoid the best practice of accessing fields via method calls?

Reply:
> the best practice of accessing fields via method calls

For external users of a class I agree, however not if there is inheritance. We extend an `abstract WrapperStore` here and thus, this class is a `WrapperStore` itself -- thus it should have direct access to the store that it wraps IMHO. At least, that is my understanding of inheritance. (Also, it's an additional unnecessary method call --- even if this is a weak argument as it's micro-optimization and might not make any practical difference.)

Reply:
I think I see where you're coming from, that if (eg) a metered state store _is a_ wrapped state store, there there should be no boundaries between them, and therefore no private state or methods on the superclass.

In practice, though, it is often useful to control the scope of members in superclasses, just as with other classes. This is why we have `private` in Java, and why subclasses cannot access `private` members on their superclasses.

For example, a common operation during code maintenence/debugging/refactoring is to find all accesses to a field. If the field is private, the task is easy, because you know that all accesses must take place in the same file. If it's protected, though, you have to consider the entire codebase. It's especially important for settable fields, but still useful for final fields.

Another benefit is referential transparency. This means that it shouldn't matter to callers whether the response is *actually* a field or computed from other members. In Scala, they tried to skirt the issue by making parens optional on getter methods. That way, you can (theoretically) switch from field to method in the future. In Java, though, you have to think ahead and set yourself up for success by preferring to access state via methods.

A concrete example of how referential transparency bit us hard is in `Windows`. By choosing to make the interface for `segments` a field, rather than a method, we made it impossible to refactor it later without breaking. When we decided to move to "segment interval" instead of "number of segments", we could have computed `segments` from the interval and retention period, but were unable to change `Windows.segments` specifically because we decided to make it a field rather than a method.

You are right that it's much more important for public APIs, but there is still value in careful engineering of private APIs. Mostly, it reduces maintenence burden.

The ultimate reason for this PR is that, in deciding to update the storage format of specific stores, we got dragged into a 10k+ LOC change that touches almost everywhere that stores appear. Generally speaking, such change amplification indicates highly coupled code. Through careful design, we would hope to be able to implement similar features in the future without a huge "splash zone".

To be clear, I'm not maligning the current state of the code... It makes sense to me to start all features in a low-fi fashion, and to employ formal "Software Engineering" judiciously when it's actually useful. In this case, we have empirical evicence that careful engineering in the store ecosystem will pay maintenence dividends in the future.

Reply:
> Another benefit is referential transparency.

If you want this, you should no inherit but encapsulate and _use_.

I think you argue for encapsulation over inheritance -- and I can buy this argument -- however, than we should not extend `WrappedStore` -- for this particular case, it seems inheritance is better than encapsulation though.

I agree that `protected` is weird in Java because there is no reason that `protected` "extends" `default / package private` visibility -- it's two independent visibility "classes" that are mangled together in Java. But we can't change Java and if we want `WrapperStore` to be extended, it seems consequently correct to me, to make it `protected`.

This is my last try to make my point -- if you insist of keeping as-is (and mangle inheritance and encapsulation together---what I don't like too much), it's ok too. (Not worth to spent more time on this.)

Reply:
Ok, it sounds like you're saying that when using inheritance, one should not have any private fields, but expose everything to the subclasses. It seems like this could be a plausible code style. I've never heard of this practice, but it might have some advantages.

I don't think I've ever heard that inheritance and encapsulation are anything but orthogonal concepts. So, whether you're using composition or inheritance, encapsulation still has the same pros and cons. As far as I'm aware, the best practice is to prefer encapsulation whenever possible. Of course, there are alternative paradigms that advocate for no private state at all, but I don't think we're talking about re-writing Streams in Lisp style ;)

I'm 100% in agreement that the access/visibility modifiers in Java make these decisions even more awkward than they need to be, but _c'est la vie_.

I also think your assessment is correct that while it's generally better to prefer composition over inheritance, this specific situation seems to dictate that the stores in question should *be* WrapperStores, and therefore should subclass it.

At the end of the day, I'm grateful that I have a clearer picture of where you were coming from with the idea that subclasses should have access to superclasses' fields, but I do still think that we should keep state private whenever possible, regardless of inheritance.